### PR TITLE
Add gdk4-win32

### PR DESCRIPTION
**Pull request by elmarco**
I could use some help, it currently fails with:

```
$ ../gir/target/release/gir -o . -m not_bound                                                                                                                        
[WARN  libgir::analysis::functions] `Win32MessageFilterFunc`: unknown type
[WARN  libgir::analysis::functions] `add_filter`: this is supposed to be a callback function but no callback was found...
[WARN  libgir::analysis::functions] `Win32MessageFilterFunc`: unknown type
[WARN  libgir::analysis::functions] `remove_filter`: this is supposed to be a callback function but no callback was found...
[NOT GENERATED METHOD] GdkWin32.Win32Display::add_filter because of GdkWin32.Win32MessageFilterFunc
[NOT GENERATED METHOD] GdkWin32.Win32Display::remove_filter because of GdkWin32.Win32MessageFilterFunc
```

and

```
error[E0433]: failed to resolve: use of undeclared crate or module `win32`
  --> gdk4-win32/src/auto/win32_hcursor.rs:25:48
   |
25 |     pub fn new(display: &Win32Display, handle: win32::HCURSOR, destroyable: bool) -> Win32HCursor {
   |                                                ^^^^^ use of undeclared crate or module `win32`

error[E0433]: failed to resolve: use of undeclared crate or module `win32`
  --> gdk4-win32/src/auto/win32_surface.rs:21:55
   |
21 |     pub fn handle(window: &impl IsA<gdk::Surface>) -> win32::HGDIOBJ {
   |                                                       ^^^^^ use of undeclared crate or module `win32`

error[E0433]: failed to resolve: use of undeclared crate or module `win32`
  --> gdk4-win32/src/auto/win32_surface.rs:28:58
   |
28 |     pub fn impl_hwnd(window: &impl IsA<gdk::Surface>) -> win32::HWND {
   |                                                          ^^^^^ use of undeclared crate or module `win32`

error[E0433]: failed to resolve: use of undeclared crate or module `win32`
  --> gdk4-win32/src/auto/win32_surface.rs:44:71
   |
44 |     pub fn lookup_for_display(display: &impl IsA<gdk::Display>, anid: win32::HWND) -> gdk::Surface {
   |                                                                       ^^^^^ use of undeclared crate or module `win32`
```

### CODE REVIEW

**Comment by elmarco** [comment]
I test the build with the pending Fedora package:
https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=2119295

```
export PKG_CONFIG_SYSROOT_DIR=/usr/x86_64-w64-mingw32/sys-root/mingw/
export PKG_CONFIG_PATH=/usr/x86_64-w64-mingw32/sys-root/mingw/lib/pkgconfig

cargo build --target x86_64-pc-windows-gnu
```

File: gdk4-win32/src/auto/win32_hcursor.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    pub fn is_destroyable(&self) -> bool {
+        glib::ObjectExt::property(self, "destroyable")
+    }
+
+    pub fn display(&self) -> Option<gdk::Display> {
+        glib::ObjectExt::property(self, "display")
+    }
+
+    //pub fn handle(&self) -> /*Unimplemented*/Basic: Pointer {
```
Comment (bilelmoussaoui) [inline gdk4-win32/src/auto/win32_hcursor.rs:52]: It seems like some property definition still needs a fixed type like what you did to others in the gir file.

File: gdk4-win32/src/functions.rs [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
```
Comment (bilelmoussaoui) [inline gdk4-win32/src/functions.rs:1]: I suppose this is not needed?

**Review by bilelmoussaoui — COMMENTED** [review]
I would also look at updating the windows job to build gdk-win so we are sure it works. I will look into reserving the name on crates.io ;)

File: gdk4-win32/Gir.toml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+version = "4.0"
+min_cfg_version = "4.0.0"
+target_path = "."
+work_mode = "normal"
+use_gi_docgen = true
+single_version_file = true
+generate_safety_asserts = true
+deprecate_by_min_version = true
+trust_return_value_nullability = true
+
```
Comment (bilelmoussaoui) [inline gdk4-win32/Gir.toml:13]: Maybe you need a external_libraries = ["Win32"] here?

File: gdk4-win32/src/auto/win32_hcursor.rs (hunk 1 above)
Comment (elmarco) [reply gdk4-win32/src/auto/win32_hcursor.rs:52]: yes, I can patch it to say it's actually a HCURSOR (which maps to isize). Gdk implementation uses g_param_spec_pointer under the hood though, so this is not satisfied:
```
54   |         glib::ObjectExt::property(self, "handle")
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `for<'b> FromValue<'b>` is not implemented for `isize`
```

I guess it has to be manually implemented then. Other ideas?

Comment (bilelmoussaoui) [reply gdk4-win32/src/auto/win32_hcursor.rs:52]: There are not that many of them, manually implementing it is fine.

File: gdk4-win32/src/auto/win32_surface.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    #[doc(alias = "gdk_win32_surface_is_win32")]
+    pub fn is_win32(window: &impl IsA<gdk::Surface>) -> bool {
+        assert_initialized_main_thread!();
+        unsafe {
+            from_glib(ffi::gdk_win32_surface_is_win32(
+                window.as_ref().to_glib_none().0,
+            ))
+        }
+    }
```
Comment (bilelmoussaoui) [inline gdk4-win32/src/auto/win32_surface.rs:41]: shouldn't those methods take a self?

Comment (elmarco) [reply gdk4-win32/src/auto/win32_surface.rs:41]: I have fixed some annotations and deprecated some functions: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4977

File: gdk4-win32/src/functions.rs (hunk 2 above)
Comment (elmarco) [reply gdk4-win32/src/functions.rs:1]: probably not needed, removed

File: gdk4-win32/Gir.toml (hunk 3 above)
Comment (elmarco) [reply gdk4-win32/Gir.toml:13]: we don't have win32 external library.. closest would be windows-rs I guess, but not really compatible with our usage. I have just defined a crate::win32. Seems to be enough, for now. We can add a feature later for windows-rs, eventually.

File: gdk4-win32/src/auto/win32_hcursor.rs (hunk 1 above)
Comment (elmarco) [reply gdk4-win32/src/auto/win32_hcursor.rs]: done. I need some solution for the filter callback stuff though... I don't get why it fails.

File: gdk4-win32/Gir.toml [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+final_type = false
+    [[object.function]]
+    name = "get_scale_factor"
+    ignore = true # docs only, otherwise it gets picked as ExtManual
+
+[[object]]
+name = "Gdk.Surface"
+status = "manual"
+final_type = false
+
```
Comment (bilelmoussaoui) [inline gdk4-win32/Gir.toml:84]: You want to add 
```
[[object]]
name = "GdkWin32.*"
status = "generate"
```
to generate the various missing functions

File: gdk4-win32/tests/abi.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
+// This file was generated by gir (https://github.com/gtk-rs/gir)
```
Comment (bilelmoussaoui) [inline gdk4-win32/tests/abi.rs:1]: I think the abi tests can't really work under windows, maybe we should look into adding an option to disable generating them or maybe they can be useful for cross compiling cases

File: gdk4-win32/Gir.toml (hunk 5 above)
Comment (elmarco) [reply gdk4-win32/Gir.toml:83]: added, although I explicitely ignore a function that I think shouldn't be exposed (win32_handle_table_lookup, see https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4980)

File: gdk4-win32/tests/abi.rs (hunk 6 above)
Comment (elmarco) [reply gdk4-win32/tests/abi.rs:1]: currently cargo test fails with:
```
User@WinDev2202Eval UCRT64 ~/src/gtk4-rs/gdk4-win32
# cargo test
  Downloaded fastrand v1.8.0
  Downloaded remove_dir_all v0.5.3
  Downloaded tempfile v3.3.0
  Downloaded gir-format-check v0.1.3
  Downloaded shell-words v1.1.0
  Downloaded 5 crates (61.9 KB) in 1.82s
   Compiling cfg-if v1.0.0
   Compiling fastrand v1.8.0
   Compiling shell-words v1.1.0
   Compiling gir-format-check v0.1.3
   Compiling remove_dir_all v0.5.3
   Compiling tempfile v3.3.0
   Compiling gdk4-win32 v0.5.0 (C:\Users\User\src\gtk4-rs\gdk4-win32)
error[E0412]: cannot find type `GdkWin32MessageFilterReturn` in this scope
   --> gdk4-win32\tests\abi.rs:206:25
    |
206 |         size: size_of::<GdkWin32MessageFilterReturn>(),
...
```

Comment (bilelmoussaoui) [reply gdk4-win32/tests/abi.rs:1]: does it work if you change https://github.com/gtk-rs/gir/blob/74b6e47217b730b04d119cdb2c7bd3d4e9de906e/src/codegen/sys/tests.rs#L510 to
 `        writeln!(w, "    (\"{ctype}\", Layout {{size: size_of::<crate::{ctype}>(), alignment: align_of::<crate::{ctype}>()}}),",`

File: gdk4-win32/src/win32_display.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+use crate::Win32Display;
+#[cfg(any(feature = "wayland_crate", feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(feature = "wayland_crate")))]
+use glib::translate::ToGlibPtr;
+#[cfg(any(all(feature = "v4_4", feature = "egl"), feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(all(feature = "v4_4", feature = "egl"))))]
+use khronos_egl as egl;
+
+impl Win32Display {
+    #[cfg(any(all(feature = "v4_4", feature = "egl"), feature = "dox"))]
```
Comment (bilelmoussaoui) [inline gdk4-win32/src/win32_display.rs:12]: the gir files doesn't seem to annotate this as being available only in v4_4?

File: gdk4-win32/src/win32_display.rs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::Win32Display;
+#[cfg(any(feature = "wayland_crate", feature = "dox"))]
```
Comment (bilelmoussaoui) [inline gdk4-win32/src/win32_display.rs:4]: you probably meant egl here

File: gdk4-win32/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+```toml
+[dependencies]
+gdk-win32= { git = "https://github.com/gtk-rs/gtk4-rs.git", package = "gdk4-win32" }
+```
+
+### Features
+
+| Feature | Description |
+| ---     | ----------- |
+| `v4_4` | Enable the new APIs part of GTK 4.4 |
```
Comment (bilelmoussaoui) [inline gdk4-win32/README.md:37]: Should include the egl feature as well

File: gdk4-win32/src/win32_display.rs (hunk 7 above)
Comment (elmarco) [reply gdk4-win32/src/win32_display.rs:12]: it's not annotated correctly,  now part of https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4980

File: gdk4-win32/src/win32_display.rs (hunk 8 above)
Comment (elmarco) [reply gdk4-win32/src/win32_display.rs:4]: yes, thanks

File: gdk4-win32/README.md (hunk 9 above)
Comment (elmarco) [reply gdk4-win32/README.md:37]: done

File: gdk4-win32/tests/abi.rs (hunk 6 above)
Comment (elmarco) [reply gdk4-win32/tests/abi.rs:1]: why would that change be necessary?

So if I get rid of `#[cfg(target_os = "linux")]` in sys/tests, it seems I can run the tests fine.

File: gdk4-win32/src/auto/win32_hcursor.rs (hunk 1 above)
Comment (elmarco) [reply gdk4-win32/src/auto/win32_hcursor.rs]: I left the add_filter/remove_filter as TODO.

File: gdk4-win32/tests/abi.rs (hunk 6 above)
Comment (elmarco) [reply gdk4-win32/tests/abi.rs:1]: @bilelmoussaoui the trouble is more about the Gdk prefix.. no idea why it is not stripped

Comment (bilelmoussaoui) [reply gdk4-win32/tests/abi.rs:1]: > @bilelmoussaoui the trouble is more about the Gdk prefix.. no idea why it is not stripped

I can't understand why they need to be stripped?

File: gdk4-win32/README.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
```
Comment (bilelmoussaoui) [inline gdk4-win32/README.md:37]: The egl/win32 features are missing

File: gdk4-win32/tests/abi.rs (hunk 6 above)
Comment (elmarco) [reply gdk4-win32/tests/abi.rs:1]: cargo test --target x86_64-pc-windows-gnu:

```
error[E0412]: cannot find type `GdkWin32MessageFilterReturn` in this scope
   --> gdk4-win32/tests/abi.rs:206:25
    |
206 |         size: size_of::<GdkWin32MessageFilterReturn>(),
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
   ::: /home/elmarco/src/gtk4-rs/gdk4-win32/src/auto/enums.rs:11:1
    |
11  | pub enum Win32MessageFilterReturn {
    | --------------------------------- similarly named enum `Win32MessageFilterReturn` defined here
    |
help: an enum with a similar name exists
    |
206 |         size: size_of::<Win32MessageFilterReturn>(),
    |                         ~~~~~~~~~~~~~~~~~~~~~~~~
help: consider importing one of these items
    |
5   | use crate::ffi::GdkWin32MessageFilterReturn;
    |
5   | use ffi::GdkWin32MessageFilterReturn;

```

**Comment by bilelmoussaoui** [comment]
Would you mind changing the windows CI to re-build everytime the gdk-win32 is changed, see https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/windows-msvc.yml#L6-L18, as well as adding a step here https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/windows-msvc.yml#L91-L95 for build, test & clippy gdk-win32

File: gdk4-win32/tests/abi.rs (hunk 6 above)
Comment (bilelmoussaoui) [reply gdk4-win32/tests/abi.rs:1]: I suppose we can land this PR and figure out the -sys tests later, we don't have working tests for gdk-x11 neither for example; see https://github.com/gtk-rs/gtk4-rs/issues/39

cc @sdroege @GuillaumeGomez, would you have an idea why the error above is happening?

**Comment by bilelmoussaoui** [comment]
> Would you mind changing the windows CI to re-build everytime the gdk-win32 is changed, see https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/windows-msvc.yml#L6-L18, as well as adding a step here https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/windows-msvc.yml#L91-L95 for build, test & clippy gdk-win32

I forgot that we should do the same for the docs job, especially the part in https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/docs.yml#L70 otherwise we won't get any gdk-win32 docs.

File: .github/workflows/docs.yml [hunk 11]
```diff
@@ -67,7 +67,15 @@ jobs:
       - uses: actions-rs/cargo@v1
         with:
           command: doc
-          args: -p gdk4 -p gdk4-sys -p gdk4-wayland -p gdk4-wayland-sys -p gdk4-x11 -p gdk4-x11-sys -p gsk4 -p gsk4-sys -p gtk4 -p gtk4-sys -p gtk4-macros --features "dox" --no-deps
+          args: >
+            -p gdk4 -p gdk4-sys
```
Comment (bilelmoussaoui) [inline .github/workflows/docs.yml:71]: It needs changes to the on push/pull folders as well

File: gdk4-win32/tests/abi.rs (hunk 6 above)
Comment (sdroege) [reply gdk4-win32/tests/abi.rs:1]: Why are the sys tests in `tests/abi.rs` instead of `sys/tests/abi.rs`?

Comment (elmarco) [reply gdk4-win32/tests/abi.rs]: good catch, no idea! now cargo test works :)

**Comment by bilelmoussaoui** [comment]
One last CI change is needed: https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/CI.yml#L5-L23 & https://github.com/gtk-rs/gtk4-rs/blob/master/.github/workflows/CI.yml#L145-L151 and it is ready to go

**Comment by bilelmoussaoui** [comment]
Sorry for always bringing yet another change to do, but the main readme at https://github.com/gtk-rs/gtk4-rs#gtk4-rs- mentions the various supported gdk backends, would it be possible to add a line for gdk win32? thanks

**Pull request merged**